### ISSUE TITLE
Setting the right client side connection timeout parameters

### ISSUE THREAD

**Issue by prateekmukhedkar** [airbytehq/airbyte#27123]
Setting the right client side connection timeout parameters when we are reading a JDBC ResultSet so that we control the timeouts.

**Comment by prateekmukhedkar** [comment]
Refer https://learn.microsoft.com/en-us/sql/connect/jdbc/connection-resiliency?view=sql-server-ver16

**Comment by nguyenaiden** [comment]
Regarding conversations around DB level retries, JDBC only provides a retry to establish an initial connection but does not provide a way to fix a broken one. This is the correct and logical behavior. For example, during a query, result sets are done in a transaction which either gets committed or rolled back. If a connection breaks during this time and even if we are able to reconnect, the result set would have already been  flushed. We will rely on the cursor and checkpointing that platform provides so that we can re-issue a new query that picks up at the last known cursor. 

This is my take on this. cc: @subodh1810, @akashkulk, @rodireich , @prateekmukhedkar. Please feel free to chime in if you have any questions/concerns.

As for timeouts, it's separated into server-side and client-side parameters

**Server-side parameters** 
`statement_timeout` (type: integer, default units: ms, default value: 0 (indefinite)): The amount of time that the query is allowed to run before aborted by the PostgreSQL server. This can be set as a connection argument to the jdbc url. e.g: `jdbc:postgresql://localhost:5432/postgres?options=-c%20statement_timeout=90000`. 12 Hours is a good value to start since even for 1TB syncs. If a query hasn't finished in 12 hours, something is most likely going wrong. We can tweak this to be lower through test syncs of large tables (>500Gb) to see the median time it takes for the query to finish. 

`lock_timeout` (type: integer, default units: ms, default value: 0 (indefinite)): The amount of time that a statement can wait for a lock before being aborted by the PostgreSQL server. This time limit applies separately to each lock acquisition attempt.  The limit applies both to explicit locking requests (such as LOCK TABLE, or SELECT FOR UPDATE without NOWAIT) and to implicitly-acquired locks. `NOTE`: this is only available for version > 9.3. This has to be directly set as a query inside of a transaction or added to the server postgresql.conf file. 

`idle_in_transaction_session_timeout` (type: integer, default units: ms, default value: 0 (indefinite)): Terminate any session that has been idle (that is, waiting for a client query) within an open transaction for longer than the specified amount of time. e.g: When a `BEGIN` statement has been run, but no queries are being executed. Can be set as a connection options parameter. e.g: `jdbc:postgresql://localhost:5432/postgres?options=-c%20idle_in_transaction_session_timeout=90000`. 

`tcp_keepalives_idle`, `tcp_keepalives_interval`, `tcp_keepalives_count`: These parameters control TCP keep-alive behavior at the operating system level. They allow the server to detect broken or idle client connections and close them accordingly. These are OS level parameters that don't affect long running queries.They only assist in identifying potential broken connections.
**Default values:**
_Linux:_
`tcp_keepalives_idle`: 7200 seconds (2 hours).
`tcp_keepalives_interval`: 75 seconds.
`tcp_keepalives_count`: 9

_Windows:_
`tcp_keepalives_idle`: 2 hours
`tcp_keepalives_interval`: 1 second
`tcp_keepalives_count`: 10

**Client-side parameters:** 
These can all be set as a query parameter in the JDBC connection URL. See example shown below:

`socketTimeout` (type: integer, default units: seconds, max value: 2147484, default value: 0 (indefinite)): The timeout value used for socket read operations. If reading from the server takes longer than this value, the connection is closed. This can be used as both a brute force global query timeout and a method of detecting network problems. This could be set as a standard connection parameter in the jdbc connection url. e.g: `jdbc:postgresql://localhost/test?user=fred&password=secret&socketTimeout=150000`. I believe this should be left untouched and offload the timeout to either statement_timeout on the server side or queryTimeout on the client side.

`connectTimeout` (type: integer, default units: seconds, max value: 2147484, default value: 10 seconds)
The timeout value used for socket connect operations. If connecting to the server takes longer than this value, the connection is broken. The timeout is specified in seconds max(2147484) and a value of zero means that it is disabled. This should be left at default since this does not affect querying of the DB. 

`loginTimeout` (type: integer, default units: seconds, max value: 2147484, default value: 0 (indefinite)): Specify how long to wait for establishment of a database connection. Similar to connectTimeout, this should be left at default since it doesn't pertain to DB querying.

`queryTimeout` (type: integer, default units: seconds, default value: 0 (indefinite)):  Sets the maximum time the JDBC driver will wait for a statement to execute before canceling it and throwing a timeout exception. This could be set by calling the `.setQueryTimeout()` method on the [PreparedStatement instantiation](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java#L357). This could be offloaded to the `statement_timeout` param instead.

**Comment by prateekmukhedkar** [comment]
@nguyenaiden @rodireich  can we try setting the client side parameters identified above in performance harness and letting it run on the 2b_users dataset to see if it makes any improvement?

**Comment by rodireich** [comment]
Regarding `statement_timeout`: on cloud our current read rate is about 4 hours for 200GB, which puts us in theory at about 20 hours for 1TB.

**Comment by subodh1810** [comment]
Can we come up with a query to log the values of the server side parameters at the beginning of the sync. Just log and nothing else changes. Also make sure that if we get an exception while trying to query these parameters, just swallow the exception and the sync should not fail

**Comment by rodireich** [comment]
There's the server config you can log, but then there's a effective setting on a connection.
In the example above, setting `options=…` on the jdbc url only affects this current connection.
If we log this configuration (which will be very helpful), it should reflect effective settings.

**Comment by prateekmukhedkar** [comment]
This connection using CDC synced 1.3 TB of data in initial snapshot https://cloud.airbyte.com/workspaces/d4772ab9-023a-49c8-8e43-aff8b0e46dc6/connections/d9bad086-a550-406d-97f3-86bf0abfa040/job-history

What JDBC connection parameters does CDC set?

**Comment by nguyenaiden** [comment]
I've compiled my findings thus far [here](https://docs.google.com/document/d/19OyLzdGyQKdXFFwF0P99OUW1RML8SEg4RpV8FTkEgCo/edit)
As of now, we've tried setting `idle_in_transaction_session_timeout`, `statement_timeout`, and Hikari `validationTimeout`. These have shown no material difference in our results. 

Plans moving forward:
1) We can mimic the exact JDBC connections of the successful 1.3TB sync a customer was able to sync with CDC 
2) Shift our focus to investigating why `Resultset.next()` is hanging. The first thing to do is enable GC logging on the pod along with running some kind of system metrics collection like visualVM to track the pod health as it handles the sync to see what's going on.
